usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
+        TRACE_DEVEL("ber_decode_PrivateKeyInfo failed\n");
+        return rc;
+    }
+
+    if (memcmp(alg, ber_AlgIdDilithium, ber_AlgIdDilithiumLen) != 0) {
+        // probably ought to use a different error
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Decode private Dilithium key */
+    rc = ber_decode_SEQUENCE(dilithium_priv_key, &buf, &buf_len, &field_len);
+    if (rc != CKR_OK)
+        return rc;
+
+    /* Now parse the key components */
+    offset = 0;
+
+    /* Version */
+    rc = ber_decode_INTEGER(buf + offset, &tmp, &len, &field_len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_INTEGER failed\n");
+        goto cleanup;
+    }
+    offset += field_len;
+
+    /* Check if key given as opaque blob */
+    if (isopaque) {
+        // opaque attribute, the Dilithium key
+        rc = ber_decode_OCTET_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_OCTET_STRING failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+    } else {
+        /* rho */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (rho) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* seed */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (seed) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* tr */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (tr) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* s1 */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (s1) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* s2 */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (s2) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* t0 */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (t0) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* t1 */
+        rc = ber_decode_BIT_STRING(buf + offset, &tmp, &len, &field_len);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_decode_BIT_STRING of (t1) failed\n");
+            goto cleanup;
+        }
+        offset += field_len;
+
+        /* Check if buffer big enough */
```

Comment:
So you do all this just to check if the individual fields fit into the sequence? 
Wouldn't it be easier to create the attributes right away, and add the check at the end? You free all the attributes already in case of an error. That way you would not need to decode the fields twice.

Reply:
Fixed. 
Let's remember to also fix ber_decode_RSAPrivateKey, ber_decode_DSAPrivateKey, and ber_decode_DHPrivateKey later.